### ISSUE TITLE
Parametric classes usage

### ISSUE THREAD

**Issue by PhilipVinc** [wesselb/plum#9]
Hi,
I would like to use parametric classes.
I'd like to implement something akin to (Julia)
```julia
struct DoubledSpace{T} where T
   space::T
end

double_int = DoubledSpace(1)
double_float = DoubledSpace(1.0)

# and dispatch on the parameter
size(a::DoubledSpace{Int}) = ...
size(a::DoubledSpace{Float64}) = ...
```

(The parameter class will be a custom class, but anyhow...)
I don't care about dispatch cost as all this is inside jax jitted function.

Anyhow, I tried to implement this by doing

```python
from plum impor dispatch, parametric

@parametric
class doubledspace:
   def __init__(self, space):
      print("initializing doubledspace")
      self.space = space

a = doubledspace(1.0)
```

but the init never gets called.

I also tried to dive into your code in `plum/parametric.py` and instrument it a bit:
```python
def parametric(Class):
    """A decorator for parametric classes."""
    subclasses = {}

    def __new__(cls, *ps):
        print(f"executing {cls.__name__}.__new__(*{ps})")
        # Only create new subclass if it doesn't exist already.
        print(f"  SubClasses are:")
        for p in subclasses:
            print(f"  \t{p}")
        print("  ---------------")
        if ps not in subclasses:
            print(f"  it is not in sublcasses")

            def __new__(cls, *args, **kw_args):
                print(f"    Calling the original {Class.__name__}.__new__(*{args}, **{kwargs}) but dropping them")
                return Class.__new__(cls)

            # Create subclass.
            name = Class.__name__ + "[" + ",".join(str(p) for p in ps) + "]"
            SubClass = type.__new__(
                CovariantMeta,
                name,
                (ParametricClass,),
                {"__new__": __new__, "_is_parametric": True},
            )
            SubClass._type_parameter = ps[0] if len(ps) == 1 else ps
            SubClass.__module__ = Class.__module__

            print(f"  Creating a new subclass called {name}")
            # Attempt to correct docstring.
            try:
                SubClass.__doc__ = Class.__doc__
            except AttributeError:  # pragma: no cover
                pass

            subclasses[ps] = SubClass

        print(f"  returning the subclass {subclasses[ps]}")
        return subclasses[ps]
```

and the output of the snippet before is
```python
➜ python -i test.py                    
now constructing
executing doubledspace.__new__(*(1,))
  SubClasses are:
  ---------------
  it is not in sublcasses
  Creating a new subclass called doubledspace[1]
  returning the subclass <class '__main__.doubledspace[1]'>
```

So it seems that even the new method
```python
            def __new__(cls, *args, **kw_args):
                print(f"    Calling the original {Class.__name__}.__new__(*{args}, **{kwargs}) but dropping them")
                return Class.__new__(cls)
```
is never called. As well as my custom init.

Can you give me a few pointers on what i should do?

**Comment by PhilipVinc** [comment]
Also if my parametric class has methods attached to it or properties, those won't work

```python
@parametric
class prova:
  a: int 
  b: int = 2
  def __init__(self, a,b=2):
  	self.a = a
  	self.b=b
  	print(f"executing init a={a}, b={b}")

  @property
  def prop(self):
    return 12

  def me(self):
  	return 10

a = prova(1)
>>> print(a.prop)
<property object at 0x7fc530cddcc0>
# should be a number

>>> print(a.me())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: me() missing 1 required positional argument: 'self'
# self is not passed. why?
```

**Comment by PhilipVinc** [comment]
Ah I see.
The point is that you should first be creating the paramtetrized type by calling `T = Test[parameters]` and then construct an object from type type `T(args...)`.

I think it would be cool to make parametrized types more ergonomic so that, `getitem` accessor is used to refer to it as a parametric type and returns the parametrized type, and the `__call__` accessor is used to automatically get the parametrized type and then construct it. 
But one would need to chose a convention.
So it's not that simple...

Somehow like this:
```python
class TypeMeta(abc.ABCMeta):
    """Types can also be instantiated with indexing."""

    def __call__(self, *args, **kwargs):
        T = self.__create_type__(*args)
        return T(*args, **kwargs)

    def __create_type__(self, *args):
        rv = self.__new__(self, *args)  # Because __new__ is static!
        if isinstance(rv, self):
            rv.__init__(*args)
        return rv

    def __getitem__(self, p):
        if isinstance(p, tuple):
            return self.__create_type__(*p)
        else:
            return self.__create_type__(p)
```

**Comment by wesselb** [comment]
Hey @PhilipVinc!

The current implementation of "parametric classes" do not allow you to construct the class without specifying the "type parameter". For example, this is how `doubledspace` would work:

```python
In [1]: from plum import dispatch, parametric
   ...:
   ...: @parametric
   ...: class doubledspace:
   ...:    def __init__(self, space):
   ...:       print("initializing doubledspace")
   ...:       self.space = space
   ...:

In [2]: a1 = doubledspace[float](1.0)
initializing doubledspace

In [3]: a1
Out[3]: <__main__.doubledspace[<class 'float'>] at 0x7f9cfb7c1390>

In [4]: a2 = doubledspace[int](1)
initializing doubledspace

In [5]: a2
Out[5]: <__main__.doubledspace[<class 'int'>] at 0x7f9cfb863c18>
```

You can then dispatch on `doublespace[int]` and `doublespace[float]`:

```python
from numbers import Number

def f(x: doubledspace[Number]):
    print("fallback")

def f(x: doubledspace[int]):
    print("int")
```

```python
In [11]: f(a1)
fallback

In [12]: f(a2)
int
```

Similarly, this is how the `prove` example would work:

```python
In [2]: from plum import dispatch, parametric

In [3]: @parametric
   ...: class prova:
   ...:   a: int
   ...:   b: int = 2
   ...:   def __init__(self, a,b=2):
   ...:     self.a = a
   ...:     self.b=b
   ...:     print(f"executing init a={a}, b={b}")
   ...:
   ...:   @property
   ...:   def prop(self):
   ...:     return 12
   ...:
   ...:   def me(self):
   ...:     return 10
   ...:

In [4]: a = prova[int](1)
executing init a=1, b=2

In [5]: print(a.me())
10
```

I have it as a TODO that you should still be able to construct the classes without specifying the type parameter, or perhaps error then.

**Comment by wesselb** [comment]
> I think it would be cool to make parametrized types more ergonomic so that, getitem accessor is used to refer to it as a parametric type and returns the parametrized type, and the __call__ accessor is used to automatically get the parametrized type and then construct it.
> But one would need to chose a convention.
> So it's not that simple...

> Somehow like this: <snip>

Ah, automatically inferring the types from the arguments that it's given is interesting. Maybe that's a reasonable default behaviour if the type parameter isn't specified. Clever!

**Comment by wesselb** [comment]
This was fixed in #11.
